### PR TITLE
WIP: iconv/libiconv improvements

### PR DESCRIPTION
**Pull request by greenc-FNAL**
First pass at a working set of `iconv`/`libintl` implementation-agnostic
packages.

Not yet vetted for libintl quirks.

### CODE REVIEW

File: var/spack/repos/builtin/packages/lftp/package.py [hunk 1]
```diff
@@ -26,11 +26,13 @@ class Lftp(AutotoolsPackage):
...
-        return [
+        args = [
             "--with-expat={0}".format(self.spec["expat"].prefix),
-            "--with-libiconv={0}".format(self.spec["iconv"].prefix),
             "--with-openssl={0}".format(self.spec["openssl"].prefix),
             "--with-readline={0}".format(self.spec["readline"].prefix),
             "--with-zlib={0}".format(self.spec["zlib"].prefix),
             "--disable-dependency-tracking",
         ]
+        if self.spec["iconv"].name == "lib":
```
Comment (marcmengel) [inline var/spack/repos/builtin/packages/lftp/package.py]: This seems oddly different from the others, shouldn't this be ... != "libc":

File: var/spack/repos/builtin/packages/neovim/package.py [hunk 2]
```diff
@@ -122,7 +122,7 @@ class Neovim(CMakePackage):
     with when("@0.6:"):
         depends_on("cmake@3.10:")
         depends_on("gperf@3.1:")
-        depends_on("libiconv@1.15:")
+        depends_on("iconv")
```
Comment (marcmengel) [inline var/spack/repos/builtin/packages/neovim/package.py]: This one worries me slightly, it loses the minimum version on libiconv if we were to use it...
maybe we could add a conflicts("libiconv@:1.14") or some such?

**Review by marcmengel — COMMENTED** [review]
Overall looks good, if you could take a look at the two comments I made in the changes...

File: var/spack/repos/builtin/packages/lftp/package.py (hunk 1 above)
Comment (greenc-FNAL) [reply var/spack/repos/builtin/packages/lftp/package.py]: Fixed.

File: var/spack/repos/builtin/packages/neovim/package.py (hunk 2 above)
Comment (greenc-FNAL) [reply var/spack/repos/builtin/packages/neovim/package.py]: Fixed (I think).

File: var/spack/repos/builtin/packages/lftp/package.py (hunk 1 above)
Comment (greenc-FNAL) [reply var/spack/repos/builtin/packages/lftp/package.py]: Fixed.

File: var/spack/repos/builtin/packages/neovim/package.py (hunk 2 above)
Comment (greenc-FNAL) [reply var/spack/repos/builtin/packages/neovim/package.py]: Fixed (I think).

**Review by marcmengel — APPROVED** [review]
That looks good now.

**Comment by greenc-FNAL** [comment]
Still not sure about `libintl`, which is why it's still a draft.

**Comment by marcmengel** [comment]
The only thing I know to get libintl right is to check` if self.spec['gettext'].prefix = '/usr' and self.spec.os in[ 'scientific7','centos7'...]` and fix the flag;  but we don't want to put that in all the packages
that mention -lintl.  (I tried a version of that about 2 years ago, they said, basically "Yecch!").

They have been discussing a scheme where you ask the package where its libraries are, example is Python, you make a:
```
    @property
    def libs(self):
           ...
           return LibraryList( path_to_library )
```
We could add that property to gettext, and put the ugliness there, in one place, and change
everyone who mentions "-lintl" (there are about 10?, 15?) explicitly to instead use
spec['gettext'].libs.  I think that change would even be accepted upstream.

**Comment by marcmengel** [comment]
See the discussion in https://spack.readthedocs.io/en/latest/packaging_guide.html#blas-lapack-and-scalapack-libraries about using libs properties.

**Comment by marcmengel** [comment]
See #36 for -lintl stuff.  I think I have the right mix there now.

**Pull request merged**